components/container_id/container_id.c

```diff
@@ -1,101 +1,266 @@
+ *
+ * Docker
+ * 13:name=systemd:/docker/3726184226f5d3147c25fdeab5b60097e378e8a720503a5e19ecfdf29f869860
+ *
+ * Kubernetes
+ * 11:perf_event:/kubepods/something/pod3d274242-8ee0-11e9-a8a6-1e68d864ef1a/3e74d3fd9db4c9dd921ae05c2502fb984d0cde1b36e581b13f79c639da4518a1
+ *
+ * ECS
+ * 9:perf_event:/ecs/user-ecs-classic/5a0d5ceddf6c44c1928d367a815d890f/38fac3e99302b3622be089dd41e7ccf38aff368a86cc339972075136ee2710ce
+ *
+ * Fargate < 1.4
+ * 11:something:/ecs/5a081c13-b8cf-4801-b427-f4601742204d/432624d2150b349fe35ba397284dea788c2bf66b885d14dfc1569b01890ca7da
+ */
+#define CONTAINER_REGEX "[0-9a-f]\\{64\\}"  // Original ERE: "[0-9a-f]{64}"
+/* Example of a valid task ID.
+ *
+ * Fargate 1.4+
+ * 1:name=systemd:/ecs/34dc0b5e626f2c5c4c5170e34b10e765-1234567890
+ */
+#define TASK_REGEX "[0-9a-f]\\{32\\}-[0-9]\\{1,20\\}"  // Original ERE: "[0-9a-f]{32}-[0-9]+"
+
+#define MIN_ID_LEN DATADOG_PHP_CONTAINER_ID_MIN_LEN
+#define MAX_ID_LEN DATADOG_PHP_CONTAINER_ID_MAX_LEN
+
+typedef datadog_php_container_id_parser dd_parser;
+
+static bool dd_parser_is_valid_line(dd_parser *parser, const char *line) {
+    return regexec(&parser->line_regex, line, 0, NULL, 0) == 0;
 }
 
-static void dd_extract_id(char *buf, char *pos) {
-    char *end = NULL;
-    size_t len = dd_str_seek_end(pos, &end);
+#define LEN_SO_FAR (end - start)
 
-    if (len < DATADOG_PHP_CONTAINER_ID_MAX_LEN) {
-        return;
-    }
+#define TASK_ID_MIN_LEN (32 + 1 + 1)   // [0-9a-f]{32}-[0-9]{1}
+#define TASK_ID_MAX_LEN (32 + 1 + 20)  // [0-9a-f]{32}-[0-9]{20}
+
+static bool dd_parser_extract_task_id(dd_parser *parser, char *buf, const char *line) {
+    if (regexec(&parser->task_regex, line, 0, NULL, 0) != 0) return false;
+
+    /* Normally we would just use 'regmatch_t' for position matching and
+     * extract the desired string from the matched position. But
+     * unfortunately if Oniguruma <= 6.9.4 is linked in, the POSIX regex
+     * symbols are overridden with Oniguruma flavored ones and 'regmatch_t'
+     * will be mangled so we cannot use it here.
+     *
+     * Ideally we would fall back to extracting the IDs using sscanf(), but
+     * since there is no format directive for a minimum or exact field width,
+     * sscanf() will often pull out other parts of the cgroup line that are
+     * not part of the target ID.
+     *
+     * That leaves us with our final old-school fallback of traversing the
+     * string one character at a time to find start and end of the target ID.
+     */
+    char *start;
+    char *end;
+    size_t len;
+
+    start = end = (char *)line;
+    len = strlen(line);
+
+    /* Traverse the string to find a task ID with the following pattern:
+     *
+     * [0-9a-f]{32}-[0-9]{1,20}
+     *
+     */
+    while ((size_t)(start - line + TASK_ID_MIN_LEN) <= len) {
+        end = start;
+
+        /* We start off looking for 32 hex chars in a row: [0-9a-f]{32} */
+        while (isxdigit(end[0]) && LEN_SO_FAR < 32) {
```

Comment:
It might be sixes, but this sort of style ends up being cleaner in some cases:

```c
char *current = start;
char *end = start + 32;
for (; current != end; ++current) {
  if (!isxdigit(*current)) break;
}
if (current != end) {
  continue;
}
```

I don't like the `LEN_SO_FAR` macro that uses local variables "magically", which is why I suggest trying it the other style.

Reply:
Resolved by #1194.